### PR TITLE
Validate petition form

### PR DESCRIPTION
**Pull request by megcrenshaw**
This is the validation for the petition signup. Still need to actually hit the endpoint and redirect on submit. The goal is to mimic https://code.org/congrats/course1.

See (new, as of afternoon of April 25) functionality here: https://watch.screencastify.com/v/z8ZEhI4Bd53kWNIZoGYf

## Testing story

My current plan is to add testing in a different PR to cover (a) the correct text shows, (b) the request is working as expected, and (c) the validation is working as expected. Happy for pushback on this––it's not currently tested, and we aren't testing most of the congrats page stuff from what I can see. @davidsbailey thoughts on this?

## Follow-up work

Need to do the following:
1) Handle submit and redirect
2) Add internalization strings––can start this as soon as we finalize the error strings
3) Add test coverage

When all is complete, I can (a) remove internalization strings from the pegasus file, and (b) remove the pegasus petition stylesheet

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/templates/certificates/petition/PetitionHelpers.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  let errorStrings = [];
+  forEach(data, (value, key) => {
+    if (keyValidation[key] && !keyValidation[key].isValid(value)) {
+      errorStrings.push(keyValidation[key].errorText);
+    }
+  });
+
+  if (errorStrings.length === 0) {
```
Comment (megcrenshaw) [inline apps/src/templates/certificates/petition/PetitionHelpers.js]: This approach really doesn't work for translation. I'm talking with Tess to figure out a good solution––I think either we (a) translate all 7 possible error strings (3 fields, either right or wrong, minus the one possibility that all fields are valid), or (b) have a general error message and highlight the invalid fields. (b) is easier but it deviates from what's currently on production

Comment (davidsbailey) [reply apps/src/templates/certificates/petition/PetitionHelpers.js]: hey Meg - my top level thought here is to mimic how we do error handling in other places in dashboard, including places where we support translations -- you do not need to replicate what you see on code.org today exactly.

Comment (davidsbailey) [reply apps/src/templates/certificates/petition/PetitionHelpers.js]: to your point, I think it is going to be necessary to give individual error messages.

File: apps/src/templates/certificates/petition/PetitionForm.jsx [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+          ['-', ...range(1, 101)].map((age, index) => (
+            <option key={index} value={age}>
+              {age}
+            </option>
+          ))
+        )}
+        {buildFieldGroup(
+          'profession',
+          'I am a',
+          <a href="/privacy">See our privacy practices for children</a>,
```
Comment (tim-dot-org) [inline apps/src/templates/certificates/petition/PetitionForm.jsx]: it looks like we're including this text twice (looks a little weird on the page)

File: apps/src/templates/certificates/petition/PetitionHelpers.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
+import React from 'react';
+import {isEmail, isInt} from '@cdo/apps/util/formatValidation';
+import FieldGroup from './FieldGroup';
+import {forEach, has} from 'lodash';
+
+export const buildControlledFieldGroup = (
```
Comment (tim-dot-org) [inline apps/src/templates/certificates/petition/PetitionHelpers.js]: could this be a react component instead?

**Review by tim-dot-org — CHANGES_REQUESTED** [review]

File: apps/src/templates/certificates/petition/PetitionHelpers.js (hunk 1 above)
Comment (megcrenshaw) [reply apps/src/templates/certificates/petition/PetitionHelpers.js]: Excellent––I'll propose that to Tess

File: apps/src/templates/certificates/petition/PetitionForm.jsx (hunk 2 above)
Comment (megcrenshaw) [reply apps/src/templates/certificates/petition/PetitionForm.jsx]: Ah jeez. Got a little too copy and paste happy! Good catch

File: apps/src/templates/certificates/petition/PetitionHelpers.js (hunk 1 above)
Comment (megcrenshaw) [reply apps/src/templates/certificates/petition/PetitionHelpers.js]: Confirmed––will move to error highlight on the field, and will need fewer translations that way

File: apps/src/templates/certificates/petition/PetitionHelpers.js (hunk 3 above)
Comment (megcrenshaw) [reply apps/src/templates/certificates/petition/PetitionHelpers.js]: Brilliant. I made a ControlledFieldGroup component––not sure it's exactly what you're looking for, but it's certainly cleaner than it was. Thoughts about it now?

File: apps/src/templates/certificates/petition/PetitionForm.jsx [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+  return (
+    <>
+      <form
+        id="petition-form"
+        className="petition-form"
+        onSubmit={handleSubmit}
+      >
+        <div className={'petition-space'}>{errorMessage}</div>
+        {buildControlledFieldGroup('name', 'Name')}
```
Comment (tim-dot-org) [inline apps/src/templates/certificates/petition/PetitionForm.jsx]: Ya, those components were exactly what I was thinking! I was hoping to get rid of these 'build' functions also by doing that, tho. Imo they're a little hard to parse when they have a bunch of similar arguments without the argument names (and so putting a component in there tells us what the arguments are)

Comment (megcrenshaw) [reply apps/src/templates/certificates/petition/PetitionForm.jsx]: Ah got it. I think that'll work too––will investigate 🤓

Comment (megcrenshaw) [reply apps/src/templates/certificates/petition/PetitionForm.jsx]: Hmm, here's my problem: I need to put that component (maybe like PetitionFieldGroup) outside of PetitionForm so it isn't rendered thousands of time. But if I do that, I lose access to the handleChange function ...https://stackoverflow.com/questions/42573017/in-react-es6-why-does-the-input-field-lose-focus-after-typing-a-character

Comment (tim-dot-org) [reply apps/src/templates/certificates/petition/PetitionForm.jsx]: oh, i see! you should be able to wrap your `handleChange` and `handleSubmit` in `useCallback` to prevent them from being redefined every render

File: apps/src/templates/certificates/petition/PetitionHelpers.js (hunk 1 above)
Comment (davidsbailey) [reply apps/src/templates/certificates/petition/PetitionHelpers.js]: hi, going back to this earlier comment -- it looks from the commit names that you've done this work now. would you mind making a new recording or screenshot to show how this looks? doesn't need to prove that it's working in every scenario, I just want to see what it looks like

Comment (megcrenshaw) [reply apps/src/templates/certificates/petition/PetitionHelpers.js]: Ah, I wouldn't say it's the best UI, but it's here: https://watch.screencastify.com/v/smMuFAhewwGAJh9L8dWb. I'm going to refactor a couple other things I think

**Comment by megcrenshaw** [comment]
New functionality is here https://watch.screencastify.com/v/z8ZEhI4Bd53kWNIZoGYf. Two questions:
1) If the form is errored, a user updates the form and re-submits, it's unclear a new request was made. Transitions for the error border and/or error message, or a spinner might help with this.
2) The styling file is huge, and the Field Group stuff is also big. There is a FieldGroup component already in the repo, but it's pretty different (with labels for inputs instead of placeholders). Could refactor some of this, but not sure it's worth it. Thoughts?

File: apps/src/templates/certificates/petition/petitionHelpers.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+};
+
+export const getErrorMessage = data => {
+  const invalidFields = getInvalidFields(data);
+  if (invalidFields.length === 0) {
+    return '';
+  } else if (invalidFields.length === 1) {
+    return 'Please fix the error below';
+  } else {
+    return 'Please fix the errors below';
```
Comment (megcrenshaw) [inline apps/src/templates/certificates/petition/petitionHelpers.js:44]: These will move to translated strings

**Review by tim-dot-org — APPROVED** [review]
re: refactoring form components, I would wait on the web component library to make any big refactors, since that will likely eventually include form components. So, as is, good to go.
The amount of styling logic is also not too concerning to me, some of it could probably be removed if we want to be looser about matching exactly the existing form, but that's the tradeoff.

File: apps/src/templates/certificates/petition/ControlledFieldGroup.jsx [hunk 6]
```diff
@@ -0,0 +1,88 @@
+import React from 'react';
+import {FormControl, FormGroup, HelpBlock} from 'react-bootstrap';
```
Comment (davidsbailey) [inline apps/src/templates/certificates/petition/ControlledFieldGroup.jsx:2]: @tim-dot-org I am curious if you have a sense of when we should be using react boostrap vs not? I think it's fine for this, but up to this point AFAIK we were only really using it in PLC (now EP Tools) land. @megcrenshaw , let's check in before adding react-bootstrap in any new places (beyond this one) outside of existing PLC stuff.

**Review by davidsbailey — APPROVED** [review]
great progress, Meg! comments are not blocking.

File: apps/src/templates/certificates/petition/ControlledFieldGroup.jsx (hunk 6 above)
Comment (tim-dot-org) [reply apps/src/templates/certificates/petition/ControlledFieldGroup.jsx:2]: Yeah, I would lean towards not using it in more places. We're currently on an older version of react-bootstrap that is going to have less support than the latest version. our web component library should replace most of the components we use in react-bootstrap or else layout type things are implementable using flex/grid.

Comment (tim-dot-org) [reply apps/src/templates/certificates/petition/ControlledFieldGroup.jsx:2]: but yea, agree with dave re: its fine for now, since there's not really an obvious replacement yet

Comment (megcrenshaw) [reply apps/src/templates/certificates/petition/ControlledFieldGroup.jsx:2]: Sounds good––it was a bit painful finding the old react bootstrap documentation, so it certainly wasn't ideal
